File: sdk/python/satgate/agent_client.py
"""
SatGate Agent Client - Economic Middleware for AI Agents

This module provides the SatGateAgentClient, which automatically handles:
- Admin token authentication for minting (OSS gateway)
- 402 Payment Required challenges (L402)
- Token caching and refresh
- Offline delegation for worker agents
- Budget tracking and alerts

Example usage (OSS Gateway):
    from satgate import SatGateAgentClient
    
    # With admin token (OSS mode)
    client = SatGateAgentClient(
        gateway_url="http://localhost:8080",
        admin_token="your-admin-token"
    )
    
    # Make requests through the gateway
    response = client.get("/api/data")
    
    # With Lightning wallet for L402 endpoints
    from satgate.agent_client import LNDWallet
    
    client = SatGateAgentClient(
        gateway_url="http://localhost:8080",
        admin_token="your-admin-token",
        wallet=LNDWallet(macaroon_path="~/.lnd/admin.macaroon")
    )
"""

import base64
import json
import os
import time
from dataclasses import dataclass, field
from typing import Any, Callable, Dict, List, Optional, Union
from urllib.parse import urljoin

import requests

from .exceptions import (
    SatGateError,
    AuthenticationError,
    BudgetExceededError,
    PaymentRequiredError,
    PaymentFailedError,
)


@dataclass
class AgentResponse:
    """Response from a SatGate-protected API call"""
    status_code: int
    headers: Dict[str, str]
    content: bytes
    cost: float = 0.0
    budget_remaining: Optional[float] = None
    
    def json(self) -> Any:
        """Parse response as JSON"""
        return json.loads(self.content)
    
    def text(self) -> str:
        """Return response as text"""
        return self.content.decode('utf-8')


@dataclass
class CachedToken:
    """A cached macaroon with metadata"""
    token: str
    signature: str
    expires_at: float
    scope: str = "api:*"
    budget_limit: Optional[float] = None
    budget_used: float = 0.0
    caveats: List[str] = field(default_factory=list)
    
    @property
    def is_expired(self) -> bool:
        return time.time() > self.expires_at
    
    @property
    def budget_remaining(self) -> Optional[float]:
        if self.budget_limit is None:
            return None
        return max(0, self.budget_limit - self.budget_used)


class LightningWallet:
    """Base class for Lightning wallets"""
    
    def pay_invoice(self, invoice: str) -> str:
        """Pay a Lightning invoice and return the preimage"""
        raise NotImplementedError
    
    def get_balance(self) -> int:
        """Return balance in satoshis"""
        raise NotImplementedError


class LNDWallet(LightningWallet):
    """LND Lightning wallet"""
    
    def __init__(
        self,
        host: str = "localhost:10009",
        macaroon_path: Optional[str] = None,
        macaroon_hex: Optional[str] = None,
        cert_path: Optional[str] = None,
        use_rest: bool = True
    ):
        self.host = host
        self.use_rest = use_rest
        
        # Load macaroon
        if macaroon_hex:
            self.macaroon = macaroon_hex
        elif macaroon_path:
            with open(macaroon_path, 'rb') as f:
                self.macaroon = f.read().hex()
        else:
            raise SatGateError("LND macaroon required")
        
        # Load cert
        self.cert_path = cert_path
    
    def pay_invoice(self, invoice: str) -> str:
        if self.use_rest:
            return self._pay_rest(invoice)
        else:
            return self._pay_grpc(invoice)
    
    def _pay_rest(self, invoice: str) -> str:
        url = f"https://{self.host}/v1/channels/transactions"
        headers = {"Grpc-Metadata-macaroon": self.macaroon}
        data = {"payment_request": invoice}
        
        resp = requests.post(
            url,
            json=data,
            headers=headers,
            verify=self.cert_path or False,
            timeout=60
        )
        
        if resp.status_code != 200:
            raise PaymentFailedError(f"LND payment failed: {resp.text}")
        
        result = resp.json()
        if result.get('payment_error'):
            raise PaymentFailedError(f"Payment error: {result['payment_error']}")
        
        return result.get('payment_preimage', '')
    
    def _pay_grpc(self, invoice: str) -> str:
        raise NotImplementedError("gRPC support requires lndgrpc package")
    
    def get_balance(self) -> int:
        url = f"https://{self.host}/v1/balance/channels"
        headers = {"Grpc-Metadata-macaroon": self.macaroon}
        
        resp = requests.get(
            url,
            headers=headers,
            verify=self.cert_path or False,
            timeout=10
        )
        
        if resp.status_code != 200:
            raise SatGateError(f"Failed to get balance: {resp.text}")
        
        return int(resp.json().get('balance', 0))


class AlbyWallet(LightningWallet):
    """Alby wallet using NWC or API"""
    
    def __init__(self, access_token: Optional[str] = None, nwc_url: Optional[str] = None):
        self.access_token = access_token or os.environ.get('ALBY_ACCESS_TOKEN')
        self.nwc_url = nwc_url or os.environ.get('ALBY_NWC_URL')
        
        if not self.access_token and not self.nwc_url:
            raise SatGateError("Alby access token or NWC URL required")
    
    def pay_invoice(self, invoice: str) -> str:
        if self.nwc_url:
            return self._pay_nwc(invoice)
        else:
            return self._pay_api(invoice)
    
    def _pay_api(self, invoice: str) -> str:
        url = "https://api.getalby.com/payments/bolt11"
        headers = {"Authorization": f"Bearer {self.access_token}"}
        data = {"invoice": invoice}
        
        resp = requests.post(url, json=data, headers=headers, timeout=60)
        
        if resp.status_code != 200:
            raise PaymentFailedError(f"Alby payment failed: {resp.text}")
        
        result = resp.json()
        return result.get('payment_preimage', '')
    
    def _pay_nwc(self, invoice: str) -> str:
        raise NotImplementedError("NWC support requires nostr package")
    
    def get_balance(self) -> int:
        url = "https://api.getalby.com/balance"
        headers = {"Authorization": f"Bearer {self.access_token}"}
        
        resp = requests.get(url, headers=headers, timeout=10)
        
        if resp.status_code != 200:
            raise SatGateError(f"Failed to get balance: {resp.text}")
        
        return int(resp.json().get('balance', 0))


class SatGateAgentClient:
    """
    SatGate Agent Client - Economic Middleware for AI Agents
    
    For the OSS gateway, uses admin token authentication to mint
    capability tokens, then uses those tokens for API access.
    
    Automatically handles:
    - Token minting via POST /api/capability/mint (X-Admin-Token)
    - Token delegation via POST /api/capability/delegate
    - L402 Payment Required challenges
    - Token caching and refresh
    - Budget tracking
    """
    
    def __init__(
        self,
        gateway_url: str,
        admin_token: Optional[str] = None,
        token: Optional[str] = None,
        wallet: Optional[LightningWallet] = None,
        budget_limit: Optional[float] = None,
        scope: str = "api:*",
        duration: str = "1h",
        timeout: int = 30,
        max_retries: int = 3,
        on_budget_alert: Optional[Callable[[float, float], None]] = None,
    ):
        """
        Initialize the agent client.
        
        Args:
            gateway_url: Base URL of the SatGate gateway
            admin_token: Admin token for minting capability tokens (X-Admin-Token header).
                         Can also be set via SATGATE_ADMIN_TOKEN environment variable.
            token: Pre-existing capability token to use (skips minting).
                   Can also be set via SATGATE_TOKEN environment variable.
            wallet: Lightning wallet for L402 payments (optional)
            budget_limit: Maximum budget to spend (optional)
            scope: Default scope for minted tokens (default: "api:*")
            duration: Default duration for minted tokens (default: "1h")
            timeout: Request timeout in seconds
            max_retries: Maximum retry attempts for 402 challenges
            on_budget_alert: Callback when budget is running low
        """
        self.gateway_url = gateway_url.rstrip('/')
        self.admin_token = admin_token or os.environ.get('SATGATE_ADMIN_TOKEN')
        self.wallet = wallet
        self.budget_limit = budget_limit
        self.default_scope = scope
        self.default_duration = duration
        self.timeout = timeout
        self.max_retries = max_retries
        self.on_budget_alert = on_budget_alert
        
        # Session for connection pooling
        self.session = requests.Session()
        
        # Token cache
        self._token_cache: Optional[CachedToken] = None
        
        # If a pre-existing token was provided, cache it
        pre_token = token or os.environ.get('SATGATE_TOKEN')
        if pre_token:
            self._token_cache = CachedToken(
                token=pre_token,
                signature="pre-provided",
                expires_at=time.time() + 3600,
                scope=scope,
            )
        
        # Total cost tracking
        self._total_cost: float = 0.0
    
    def _get_token(self) -> Optional[str]:
        """Get a valid token, minting if necessary"""
        # Check cache
        if self._token_cache and not self._token_cache.is_expired:
            return self._token_cache.token
        
        # No admin token - can't mint
        if not self.admin_token:
            return None
        
        # Mint new token via OSS endpoint
        try:
            resp = self.session.post(
                f"{self.gateway_url}/api/capability/mint",
                json={
                    "scope": self.default_scope,
                    "duration": self.default_duration,
                },
                headers={
                    "Content-Type": "application/json",
                    "X-Admin-Token": self.admin_token,
                },
                timeout=self.timeout
            )
            
            if resp.status_code != 200:
                raise AuthenticationError(f"Mint failed ({resp.status_code}): {resp.text}")
            
            result = resp.json()
            
            # Parse expiration
            expires_at = time.time() + 3600  # Default 1 hour
            if 'expiresAt' in result:
                try:
                    from datetime import datetime
                    dt = datetime.fromisoformat(result['expiresAt'].replace('Z', '+00:00'))
                    expires_at = dt.timestamp()
                except (ValueError, ImportError):
                    pass
            
            # Cache token
            self._token_cache = CachedToken(
                token=result['token'],
                signature=result.get('signature', ''),
                expires_at=expires_at,
                scope=result.get('scope', self.default_scope),
            )
            
            return self._token_cache.token
            
        except requests.RequestException as e:
            raise AuthenticationError(f"Failed to mint token: {e}")
    
    def _handle_l402(self, response: requests.Response, method: str, url: str, **kwargs) -> AgentResponse:
        """Handle an L402 Payment Required response"""
        www_auth = response.headers.get('WWW-Authenticate', '')
        
        challenge = {}
        content_type = response.headers.get('Content-Type', '')
        if 'application/json' in content_type:
            try:
                challenge = response.json()
            except Exception:
                pass
        
        if not self.wallet:
            raise PaymentRequiredError(
                "L402 payment required but no wallet configured",
                challenge=challenge
            )
        
        # Parse L402 header: L402 macaroon="...", invoice="..."
        macaroon = ""
        invoice = ""
        
        for part in www_auth.split(','):
            part = part.strip()
            if 'macaroon=' in part:
                macaroon = part.split('=', 1)[1].strip('"\'')
            elif 'invoice=' in part:
                invoice = part.split('=', 1)[1].strip('"\'')
        
        # Also check JSON challenge
        if not invoice:
            invoice = challenge.get('invoice', '')
        if not macaroon:
            macaroon = challenge.get('macaroon', '')
        
        if not invoice:
            raise PaymentRequiredError("No invoice in L402 challenge", challenge=challenge)
        
        # Pay the invoice
        try:
            preimage = self.wallet.pay_invoice(invoice)
        except Exception as e:
            raise PaymentFailedError(f"Lightning payment failed: {e}")
        
        # Build L402 token and retry
        l402_token = f"L402 {macaroon}:{preimage}"
        headers = kwargs.get('headers', {})
        headers['Authorization'] = l402_token
        kwargs['headers'] = headers
        
        retry_resp = self.session.request(method, url, timeout=self.timeout, **kwargs)
        
        cost_sats = challenge.get('amount_sats', challenge.get('amount', 0))
        
        return AgentResponse(
            status_code=retry_resp.status_code,
            headers=dict(retry_resp.headers),
            content=retry_resp.content,
            cost=cost_sats
        )
    
    def request(
        self,
        method: str,
        path: str,
        **kwargs
    ) -> AgentResponse:
        """
        Make an HTTP request with automatic token and L402 handling.
        
        Args:
            method: HTTP method (GET, POST, etc.)
            path: URL path (will be joined with gateway_url)
            **kwargs: Additional arguments passed to requests
            
        Returns:
            AgentResponse with status, content, and cost information
        """
        url = urljoin(self.gateway_url + "/", path.lstrip("/"))
        
        # Add capability token if available
        token = self._get_token()
        if token:
            headers = kwargs.get('headers', {})
            if 'Authorization' not in headers:
                headers['Authorization'] = f"Bearer {token}"
                kwargs['headers'] = headers
        
        # Make request
        for attempt in range(self.max_retries):
            resp = self.session.request(method, url, timeout=self.timeout, **kwargs)
            
            # Success
            if resp.status_code < 400:
                return AgentResponse(
                    status_code=resp.status_code,
                    headers=dict(resp.headers),
                    content=resp.content,
                    cost=0.0,
                    budget_remaining=self._token_cache.budget_remaining if self._token_cache else None
                )
            
            # Payment required (L402)
            if resp.status_code == 402:
                return self._handle_l402(resp, method, url, **kwargs)
            
            # Auth error - try refreshing token
            if resp.status_code == 401 and attempt < self.max_retries - 1:
                self._token_cache = None
                token = self._get_token()
                if token:
                    headers = kwargs.get('headers', {})
                    headers['Authorization'] = f"Bearer {token}"
                    kwargs['headers'] = headers
                continue
            
            # Other error
            raise SatGateError(f"Request failed: {resp.status_code} {resp.text}")
        
        raise SatGateError("Max retries exceeded")
    
    def get(self, path: str, **kwargs) -> AgentResponse:
        """Make a GET request"""
        return self.request("GET", path, **kwargs)
    
    def post(self, path: str, **kwargs) -> AgentResponse:
        """Make a POST request"""
        return self.request("POST", path, **kwargs)
    
    def put(self, path: str, **kwargs) -> AgentResponse:
        """Make a PUT request"""
        return self.request("PUT", path, **kwargs)
    
    def delete(self, path: str, **kwargs) -> AgentResponse:
        """Make a DELETE request"""
        return self.request("DELETE", path, **kwargs)
    
    def ping(self) -> dict:
        """
        Ping the gateway to verify the capability token is valid.
        
        Calls GET /api/capability/ping with the Bearer token.
        
        Returns:
            dict with validation result
        """
        response = self.get("/api/capability/ping")
        return response.json()
    
    def validate_token(self, token: Optional[str] = None) -> dict:
        """
        Validate a capability token.
        
        Args:
            token: Token to validate. If not provided, uses the cached token.
            
        Returns:
            dict with validation result including caveats
        """
        t = token or (self._token_cache.token if self._token_cache else None)
        if not t:
            raise SatGateError("No token available to validate")
        
        resp = self.session.post(
            f"{self.gateway_url}/api/capability/validate",
            json={"token": t},
            timeout=self.timeout
        )
        
        if resp.status_code != 200:
            raise SatGateError(f"Validation failed: {resp.text}")
        
        return resp.json()
    
    def delegate(
        self,
        caveats: Optional[List[str]] = None,
    ) -> str:
        """
        Create a delegated (child) token for a worker agent.
        
        Uses POST /api/capability/delegate with the current token as parent.
        
        Args:
            caveats: Additional caveats/restrictions to add to the child token
            
        Returns:
            The delegated child token string
        """
        if not self._token_cache:
            self._get_token()
        
        if not self._token_cache:
            raise SatGateError("No token available for delegation")
        
        resp = self.session.post(
            f"{self.gateway_url}/api/capability/delegate",
            json={
                "parentToken": self._token_cache.token,
                "caveats": caveats or [],
            },
            timeout=self.timeout
        )
        
        if resp.status_code != 200:
            raise SatGateError(f"Delegation failed: {resp.text}")
        
        return resp.json()['token']
    
    @property
    def total_cost(self) -> float:
        """Total cost incurred by this client"""
        return self._total_cost
    
    @property
    def budget_remaining(self) -> Optional[float]:
        """Remaining budget (if budget_limit was set)"""
        if self.budget_limit is None:
            return None
        return max(0, self.budget_limit - self._total_cost)
    
    @property
    def current_token(self) -> Optional[str]:
        """The currently cached capability token, if any"""
        if self._token_cache and not self._token_cache.is_expired:
            return self._token_cache.token
        return None
    
    @property
    def current_signature(self) -> Optional[str]:
        """The signature of the currently cached token, if any"""
        if self._token_cache:
            return self._token_cache.signature
        return None

File: sdk/nodejs/src/agent-client.ts
/**
 * SatGate Agent Client - Economic Middleware for AI Agents (OSS)
 * 
 * Automatically handles:
 * - Token minting via POST /api/capability/mint (X-Admin-Token)
 * - Token delegation via POST /api/capability/delegate
 * - L402 Payment Required challenges
 * - Token caching and refresh
 * - Budget tracking and alerts
 * 
 * @example
 * ```typescript
 * import { SatGateAgentClient } from '@satgate/sdk';
 * 
 * // With admin token (auto-mints capability tokens)
 * const client = new SatGateAgentClient({
 *   gatewayUrl: 'http://localhost:8080',
 *   adminToken: 'your-admin-token'
 * });
 * 
 * // Make requests - token management is automatic
 * const response = await client.get('/api/data');
 * console.log(response.data);
 * ```
 */

import * as fs from 'fs';
import {
  SatGateError,
  AuthenticationError,
  PaymentRequiredError,
  PaymentFailedError,
  BudgetExceededError,
} from './errors';

export interface AgentResponse<T = unknown> {
  statusCode: number;
  headers: Record<string, string>;
  data: T;
  cost: number;
  budgetRemaining?: number;
}

export interface CachedToken {
  token: string;
  signature: string;
  expiresAt: number;
  scope: string;
  budgetLimit?: number;
  budgetUsed: number;
  caveats: string[];
}

// Lightning Wallets

export interface LightningWallet {
  payInvoice(invoice: string): Promise<string>;
  getBalance(): Promise<number>;
}

export interface LNDWalletOptions {
  host?: string;
  macaroonPath?: string;
  macaroonHex?: string;
  certPath?: string;
}

export class LNDWallet implements LightningWallet {
  private host: string;
  private macaroon: string;
  private certPath?: string;
  
  constructor(options: LNDWalletOptions) {
    this.host = options.host || 'localhost:8080';
    this.certPath = options.certPath;
    
    if (options.macaroonHex) {
      this.macaroon = options.macaroonHex;
    } else if (options.macaroonPath) {
      const data = fs.readFileSync(options.macaroonPath);
      this.macaroon = data.toString('hex');
    } else {
      throw new SatGateError('LND macaroon required');
    }
  }
  
  async payInvoice(invoice: string): Promise<string> {
    const url = `https://${this.host}/v1/channels/transactions`;
    
    const response = await fetch(url, {
      method: 'POST',
      headers: {
        'Grpc-Metadata-macaroon': this.macaroon,
        'Content-Type': 'application/json',
      },
      body: JSON.stringify({ payment_request: invoice }),
    });
    
    if (!response.ok) {
      throw new PaymentFailedError(`LND payment failed: ${await response.text()}`);
    }
    
    const result = await response.json() as { payment_preimage?: string; payment_error?: string };
    
    if (result.payment_error) {
      throw new PaymentFailedError(`Payment error: ${result.payment_error}`);
    }
    
    return result.payment_preimage || '';
  }
  
  async getBalance(): Promise<number> {
    const url = `https://${this.host}/v1/balance/channels`;
    
    const response = await fetch(url, {
      headers: { 'Grpc-Metadata-macaroon': this.macaroon },
    });
    
    if (!response.ok) {
      throw new SatGateError(`Failed to get balance: ${await response.text()}`);
    }
    
    const result = await response.json() as { balance?: string };
    return parseInt(result.balance || '0', 10);
  }
}

export interface AlbyWalletOptions {
  accessToken?: string;
  nwcUrl?: string;
}

export class AlbyWallet implements LightningWallet {
  private accessToken?: string;
  private nwcUrl?: string;
  
  constructor(options: AlbyWalletOptions = {}) {
    this.accessToken = options.accessToken || process.env.ALBY_ACCESS_TOKEN;
    this.nwcUrl = options.nwcUrl || process.env.ALBY_NWC_URL;
    
    if (!this.accessToken && !this.nwcUrl) {
      throw new SatGateError('Alby access token or NWC URL required');
    }
  }
  
  async payInvoice(invoice: string): Promise<string> {
    if (!this.accessToken) {
      throw new SatGateError('Alby access token required for API payments');
    }
    
    const response = await fetch('https://api.getalby.com/payments/bolt11', {
      method: 'POST',
      headers: {
        'Authorization': `Bearer ${this.accessToken}`,
        'Content-Type': 'application/json',
      },
      body: JSON.stringify({ invoice }),
    });
    
    if (!response.ok) {
      throw new PaymentFailedError(`Alby payment failed: ${await response.text()}`);
    }
    
    const result = await response.json() as { payment_preimage?: string };
    return result.payment_preimage || '';
  }
  
  async getBalance(): Promise<number> {
    if (!this.accessToken) {
      throw new SatGateError('Alby access token required');
    }
    
    const response = await fetch('https://api.getalby.com/balance', {
      headers: { 'Authorization': `Bearer ${this.accessToken}` },
    });
    
    if (!response.ok) {
      throw new SatGateError(`Failed to get balance: ${await response.text()}`);
    }
    
    const result = await response.json() as { balance?: number };
    return result.balance || 0;
  }
}

// Agent Client

export interface SatGateAgentClientOptions {
  /** Gateway URL */
  gatewayUrl: string;
  /** 
   * Admin token for minting capability tokens (X-Admin-Token header).
   * Can also be set via SATGATE_ADMIN_TOKEN environment variable.
   */
  adminToken?: string;
  /**
   * Pre-existing capability token (alternative to adminToken).
   * Can also be set via SATGATE_TOKEN environment variable.
   */
  token?: string;
  /** Lightning wallet for L402 payments */
  wallet?: LightningWallet;
  /** Maximum budget to spend */
  budgetLimit?: number;
  /** Default scope for minted tokens (default: "api:*") */
  scope?: string;
  /** Default duration for minted tokens (default: "1h") */
  duration?: string;
  /** Request timeout in milliseconds */
  timeout?: number;
  /** Maximum retry attempts for 402 challenges */
  maxRetries?: number;
  /** Callback when budget is running low */
  onBudgetAlert?: (remaining: number, limit: number) => void;
}

export class SatGateAgentClient {
  private gatewayUrl: string;
  private adminToken?: string;
  private wallet?: LightningWallet;
  private budgetLimit?: number;
  private defaultScope: string;
  private defaultDuration: string;
  private timeout: number;
  private maxRetries: number;
  private onBudgetAlert?: (remaining: number, limit: number) => void;
  
  private tokenCache?: CachedToken;
  private totalCost: number = 0;
  
  constructor(options: SatGateAgentClientOptions) {
    this.gatewayUrl = options.gatewayUrl.replace(/\/$/, '');
    this.adminToken = options.adminToken || process.env.SATGATE_ADMIN_TOKEN;
    this.wallet = options.wallet;
    this.budgetLimit = options.budgetLimit;
    this.defaultScope = options.scope || 'api:*';
    this.defaultDuration = options.duration || '1h';
    this.timeout = options.timeout || 30000;
    this.maxRetries = options.maxRetries || 3;
    this.onBudgetAlert = options.onBudgetAlert;
    
    // If a pre-existing token was provided, cache it
    const preToken = options.token || process.env.SATGATE_TOKEN;
    if (preToken) {
      this.tokenCache = {
        token: preToken,
        signature: 'pre-provided',
        expiresAt: Date.now() + 3600000,
        scope: this.defaultScope,
        budgetUsed: 0,
        caveats: [],
      };
    }
  }
  
  private async getToken(): Promise<string | undefined> {
    // Check cache
    if (this.tokenCache && Date.now() < this.tokenCache.expiresAt) {
      return this.tokenCache.token;
    }
    
    // No admin token - can't mint
    if (!this.adminToken) {
      return undefined;
    }
    
    // Mint new token via OSS endpoint
    try {
      const controller = new AbortController();
      const timeoutId = setTimeout(() => controller.abort(), this.timeout);
      
      const response = await fetch(`${this.gatewayUrl}/api/capability/mint`, {
        method: 'POST',
        headers: {
          'Content-Type': 'application/json',
          'X-Admin-Token': this.adminToken,
        },
        body: JSON.stringify({
          scope: this.defaultScope,
          duration: this.defaultDuration,
        }),
        signal: controller.signal,
      });
      
      clearTimeout(timeoutId);
      
      if (!response.ok) {
        throw new AuthenticationError(`Mint failed (${response.status}): ${await response.text()}`);
      }
      
      const result = await response.json() as {
        token: string;
        signature: string;
        scope?: string;
        expiresAt?: string;
      };
      
      // Parse expiration
      let expiresAt = Date.now() + 3600000; // Default 1 hour
      if (result.expiresAt) {
        const parsed = new Date(result.expiresAt).getTime();
        if (!isNaN(parsed)) {
          expiresAt = parsed;
        }
      }
      
      this.tokenCache = {
        token: result.token,
        signature: result.signature || '',
        expiresAt,
        scope: result.scope || this.defaultScope,
        budgetUsed: 0,
        caveats: [],
      };
      
      return this.tokenCache.token;
    } catch (e) {
      if (e instanceof SatGateError) throw e;
      throw new AuthenticationError(`Failed to mint token: ${e}`);
    }
  }
  
  private async handleL402(
    response: Response,
    method: string,
    url: string,
    options: RequestInit
  ): Promise<AgentResponse> {
    const wwwAuth = response.headers.get('WWW-Authenticate') || '';
    
    let challenge: Record<string, unknown> = {};
    const contentType = response.headers.get('Content-Type') || '';
    if (contentType.includes('application/json')) {
      try {
        challenge = await response.json() as Record<string, unknown>;
      } catch {}
    }
    
    if (!this.wallet) {
      throw new PaymentRequiredError('L402 payment required but no wallet configured', challenge);
    }
    
    // Parse L402 header: L402 macaroon="...", invoice="..."
    let macaroon = '';
    let invoice = '';
    
    for (const part of wwwAuth.split(',')) {
      const trimmed = part.trim();
      if (trimmed.includes('macaroon=')) {
        macaroon = trimmed.split('=')[1]?.replace(/['"]/g, '') || '';
      } else if (trimmed.includes('invoice=')) {
        invoice = trimmed.split('=')[1]?.replace(/['"]/g, '') || '';
      }
    }
    
    // Also check JSON challenge
    if (!invoice) invoice = challenge.invoice as string || '';
    if (!macaroon) macaroon = challenge.macaroon as string || '';
    
    if (!invoice) {
      throw new PaymentRequiredError('No invoice in L402 challenge', challenge);
    }
    
    // Pay the invoice
    let preimage: string;
    try {
      preimage = await this.wallet.payInvoice(invoice);
    } catch (e) {
      throw new PaymentFailedError(`Lightning payment failed: ${e}`);
    }
    
    // Build L402 token and retry
    const l402Token = `L402 ${macaroon}:${preimage}`;
    const headers = new Headers(options.headers);
    headers.set('Authorization', l402Token);
    
    const retryResponse = await fetch(url, { ...options, headers });
    const costSats = (challenge.amount_sats as number) || (challenge.amount as number) || 0;
    
    return {
      statusCode: retryResponse.status,
      headers: Object.fromEntries(retryResponse.headers.entries()),
      data: await retryResponse.json(),
      cost: costSats,
    };
  }
  
  async request<T = unknown>(
    method: string,
    path: string,
    options: RequestInit = {}
  ): Promise<AgentResponse<T>> {
    const url = new URL(path, this.gatewayUrl).toString();
    
    // Add capability token if available
    const token = await this.getToken();
    const headers = new Headers(options.headers);
    if (token && !headers.has('Authorization')) {
      headers.set('Authorization', `Bearer ${token}`);
    }
    
    const requestOptions: RequestInit = {
      ...options,
      method,
      headers,
    };
    
    // Make request with retries
    for (let attempt = 0; attempt < this.maxRetries; attempt++) {
      const controller = new AbortController();
      const timeoutId = setTimeout(() => controller.abort(), this.timeout);
      
      try {
        const response = await fetch(url, {
          ...requestOptions,
          signal: controller.signal,
        });
        
        clearTimeout(timeoutId);
        
        // Success
        if (response.ok) {
          return {
            statusCode: response.status,
            headers: Object.fromEntries(response.headers.entries()),
            data: await response.json() as T,
            cost: 0,
            budgetRemaining: this.tokenCache?.budgetLimit
              ? this.tokenCache.budgetLimit - this.tokenCache.budgetUsed
              : undefined,
          };
        }
        
        // Payment required (L402)
        if (response.status === 402) {
          return this.handleL402(response, method, url, requestOptions) as Promise<AgentResponse<T>>;
        }
        
        // Auth error - refresh token
        if (response.status === 401 && attempt < this.maxRetries - 1) {
          this.tokenCache = undefined;
          const newToken = await this.getToken();
          if (newToken) {
            headers.set('Authorization', `Bearer ${newToken}`);
          }
          continue;
        }
        
        // Other error
        throw new SatGateError(`Request failed: ${response.status} ${await response.text()}`);
        
      } catch (e) {
        clearTimeout(timeoutId);
        if (e instanceof SatGateError) throw e;
        throw new SatGateError(`Request failed: ${e}`);
      }
    }
    
    throw new SatGateError('Max retries exceeded');
  }
  
  async get<T = unknown>(path: string, options?: RequestInit): Promise<AgentResponse<T>> {
    return this.request<T>('GET', path, options);
  }
  
  async post<T = unknown>(path: string, body?: unknown, options?: RequestInit): Promise<AgentResponse<T>> {
    return this.request<T>('POST', path, {
      ...options,
      headers: {
        'Content-Type': 'application/json',
        ...Object.fromEntries(new Headers(options?.headers).entries()),
      },
      body: body ? JSON.stringify(body) : undefined,
    });
  }
  
  async put<T = unknown>(path: string, body?: unknown, options?: RequestInit): Promise<AgentResponse<T>> {
    return this.request<T>('PUT', path, {
      ...options,
      headers: {
        'Content-Type': 'application/json',
        ...Object.fromEntries(new Headers(options?.headers).entries()),
      },
      body: body ? JSON.stringify(body) : undefined,
    });
  }
  
  async delete<T = unknown>(path: string, options?: RequestInit): Promise<AgentResponse<T>> {
    return this.request<T>('DELETE', path, options);
  }
  
  /**
   * Ping the gateway to verify the capability token is valid.
   * 
   * GET /api/capability/ping with Bearer token
   */
  async ping(): Promise<Record<string, unknown>> {
    const response = await this.get<Record<string, unknown>>('/api/capability/ping');
    return response.data;
  }
  
  /**
   * Validate the current capability token.
   * 
   * POST /api/capability/validate
   */
  async validateToken(token?: string): Promise<Record<string, unknown>> {
    const t = token || this.tokenCache?.token;
    if (!t) throw new SatGateError('No token available to validate');
    
    const response = await this.post<Record<string, unknown>>(
      '/api/capability/validate',
      { token: t }
    );
    return response.data;
  }
  
  /**
   * Create a delegated (child) token for a worker agent.
   * 
   * POST /api/capability/delegate
   */
  async delegate(options: {
    caveats?: string[];
  } = {}): Promise<string> {
    if (!this.tokenCache) {
      await this.getToken();
    }
    
    if (!this.tokenCache) {
      throw new SatGateError('No token available for delegation');
    }
    
    const controller = new AbortController();
    const timeoutId = setTimeout(() => controller.abort(), this.timeout);
    
    try {
      const response = await fetch(`${this.gatewayUrl}/api/capability/delegate`, {
        method: 'POST',
        headers: { 'Content-Type': 'application/json' },
        body: JSON.stringify({
          parentToken: this.tokenCache.token,
          caveats: options.caveats || [],
        }),
        signal: controller.signal,
      });
      
      clearTimeout(timeoutId);
      
      if (!response.ok) {
        throw new SatGateError(`Delegation failed: ${await response.text()}`);
      }
      
      const result = await response.json() as { token: string };
      return result.token;
    } catch (e) {
      clearTimeout(timeoutId);
      if (e instanceof SatGateError) throw e;
      throw new SatGateError(`Delegation failed: ${e}`);
    }
  }
  
  /** Total cost incurred by this client */
  getTotalCost(): number {
    return this.totalCost;
  }
  
  /** Remaining budget (if budgetLimit was set) */
  getBudgetRemaining(): number | undefined {
    if (this.budgetLimit === undefined) return undefined;
    return Math.max(0, this.budgetLimit - this.totalCost);
  }
  
  /** The current cached token, if any */
  getCurrentToken(): string | undefined {
    if (this.tokenCache && Date.now() < this.tokenCache.expiresAt) {
      return this.tokenCache.token;
    }
    return undefined;
  }
  
  /** The signature of the current cached token, if any */
  getCurrentSignature(): string | undefined {
    return this.tokenCache?.signature;
  }
}
